### PR TITLE
#407:TailSampling: Added string tag probabilistic filter with counting…

### PR DESCRIPTION
**Pull request by vikrambe**
… sampler.  

**Usecase:**
Idea here is to capture all Traces that have errors and still capture certain percentage of traces that are successful. 

_For example : I would want to sample 100% of all http.status_code "500"s and 10% of http.status_code "200"s so on and so forth_

**Background:**
Currently, collector's tail based sampler supports sampling based on span tag value (i.e. via string_tag_filter). For example, let's say we want to sample all the traces that  have at-least one span with http.status_code=500 then above sampler is good. And it makes sense to capture all error status codes for triaging purpose. But in order to reduce load on the tracing backend, we would want to capture certain percentage (i.e. configured value) of successful trace.

**Solution**
Supported string_tag_probabilistic_filter that is an enhanced version of string_tag_filter.  It allows  adding a percentage [i.e. value between 0.0 and 1.0]  delimited by "::" along with the string tag value. When the tag value matches the sampling criteria, we also check if it falls within the counting sampler bucket and only then return the sampling decision. 

_Config Example:
{
            name: test-policy-5,
            type: string_probabilistic_attribute,
            string_probabilistic_attribute: {key: key3, values: [value1::1.0, value2::0.2]}
}_

_Sampling percentage can be given as below 
0.0 -> Dont Sample
0.1-> 10%
0.2->20%
...
0.9->90%
1.0->100%_

### CODE REVIEW

**Comment by vikrambe** [comment]
I signed it

**Comment by vikrambe** [comment]
I signed it

File: processor/samplingprocessor/tailsamplingprocessor/sampling/string_tag_probabilistic_filter.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+			v := strings.Split(value, "::")
+			valuesMap[v[0]] = true
+			f, _ := strconv.ParseFloat(v[1], 32)
+			probabilityMap[v[0]] = NewCountingSampler(float32(f))
+		}
+	}
+	return &stringProbabilisticAttributeFilter{
+		key:         key,
+		values:      valuesMap,
+		probability: probabilityMap,
```
Comment (rghetia) [inline processor/samplingprocessor/tailsamplingprocessor/sampling/string_tag_probabilistic_filter.go]: how about just using probabilityMap. If map holds the entry (sampler) then use the sampler. No need for valuesMap.

File: processor/samplingprocessor/tailsamplingprocessor/sampling/counting_sampler.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+package sampling
+
+import (
+	"math/rand"
+	"sync/atomic"
+	"time"
+)
+
+//DIVISOR : Divisor constant
+const DIVISOR = 100
```
Comment (rghetia) [inline processor/samplingprocessor/tailsamplingprocessor/sampling/counting_sampler.go]: is precision sufficient? I would think that for high QPS environment if may need precision of 0.001 if not higher.

File: processor/samplingprocessor/tailsamplingprocessor/sampling/counting_sampler_test.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	assert.Equal(t, getSampleBucketSize(sampler, DIVISOR), 20)
+
+	// Sample 50%
+	sampler = NewCountingSampler(0.5)
+	assert.Equal(t, getSampleBucketSize(sampler, DIVISOR), 50)
+
+	// Sample 100%
+	sampler = NewCountingSampler(1.0)
+	assert.Equal(t, getSampleBucketSize(sampler, DIVISOR), 100)
+}
```
Comment (rghetia) [inline processor/samplingprocessor/tailsamplingprocessor/sampling/counting_sampler_test.go]: Please add test for higher precision (0.15 or 0.25).

File: processor/samplingprocessor/tailsamplingprocessor/sampling/string_tag_probabilistic_filter_test.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+func TestEvaluate_TenPercent(t *testing.T) {
+	var status = "200"
+	policies := []string{status + "::0.1"}
+	var stringProbabilisticAttributeFilter = NewStringProbabilisticAttributeFilter(string("http.status_code"), policies)
+	j := 0
+	for i := 0; i < 100; i++ {
+		j = evaluateStringProbabilityAttrib(stringProbabilisticAttributeFilter, j, status)
+	}
+	assert.Equal(t, 10, j)
+}
```
Comment (rghetia) [inline processor/samplingprocessor/tailsamplingprocessor/sampling/string_tag_probabilistic_filter_test.go]: Please add test for 15 percent.

File: processor/samplingprocessor/tailsamplingprocessor/config.go [hunk 5]
```diff
@@ -34,6 +34,10 @@ const (
 	StringAttribute PolicyType = "string_attribute"
 	// RateLimiting allows all traces until the specified limits are satisfied.
 	RateLimiting PolicyType = "rate_limiting"
+	// StringProbabilisticAttribute sample traces that a attribute, of type string, matching
```
Comment (rghetia) [inline processor/samplingprocessor/tailsamplingprocessor/config.go]: nit: s/a attribute/an attribute/

**Comment by vikrambe** [comment]
@rghetia @tigrannajaryan thanks for the review comments. I will take care of the comments and update PR shortly

**Comment by vikrambe** [comment]
> @vikrambe thank you for working on this. Before we move forward can you please add a description to the PR and include the problem that this is solving and the use cases for the new functionality.

@tigrannajaryan Thanks. Added description to the PR.

File: processor/samplingprocessor/tailsamplingprocessor/sampling/string_tag_probabilistic_filter.go (hunk 1 above)
Comment (vikrambe) [reply processor/samplingprocessor/tailsamplingprocessor/sampling/string_tag_probabilistic_filter.go]: Good Suggestion. Taken care. Thanks

File: processor/samplingprocessor/tailsamplingprocessor/sampling/counting_sampler.go (hunk 2 above)
Comment (vikrambe) [reply processor/samplingprocessor/tailsamplingprocessor/sampling/counting_sampler.go]: Hmm. Maybe we can make this configurable. I will ensure to externalize this in my next PR.

File: processor/samplingprocessor/tailsamplingprocessor/sampling/counting_sampler_test.go (hunk 3 above)
Comment (vikrambe) [reply processor/samplingprocessor/tailsamplingprocessor/sampling/counting_sampler_test.go]: Done

File: processor/samplingprocessor/tailsamplingprocessor/sampling/string_tag_probabilistic_filter_test.go (hunk 4 above)
Comment (vikrambe) [reply processor/samplingprocessor/tailsamplingprocessor/sampling/string_tag_probabilistic_filter_test.go]: Done

File: processor/samplingprocessor/tailsamplingprocessor/config.go (hunk 5 above)
Comment (vikrambe) [reply processor/samplingprocessor/tailsamplingprocessor/config.go]: Done

**Comment by tigrannajaryan** [comment]
@vikrambe thanks for the description. I believe this overlaps with the functionality that we were intending to implement, something called composite policy for tail-based sampler. We have a draft design for the feature. Please have a look and let us know if this would cover your use case: https://docs.google.com/document/d/10wpIv3TtXgOik05smHm3nYeBX48Bj76TCMxPy8e1NZw/edit#heading=h.ecy5l2puwtp4

@pjanotti what are your thoughts on this? It seems composite policy has a real-world use case.

**Comment by vikrambe** [comment]
> https://docs.google.com/document/d/10wpIv3TtXgOik05smHm3nYeBX48Bj76TCMxPy8e1NZw/edit#heading=h.ecy5l2puwtp4

@tigrannajaryan @pjanotti @rghetia Nice, design is generic. I would like to implement this feature. If you have not started implementing it yet, let me know if you would want me to update this PR with the composite-policy Or Open a new PR ? 

w.r.t tail-sampling : Do we have a feature/design doc that solves tail-sampling across multiple otel-collector instances ?

**UseCase** : In order to handle high TPS, we would need multiple instances of otel collector behind a LoadBalancer. In this case, spans belonging to a trace are not guaranteed to land-up in the same collector instance. As the tail sampling works on the spans received by a single instance, sampling decision may potentially be flawed.

Solution : Enhance Opentelemetry collector to use distributed cache, so that OtelCollectors have a single view of the cache.

**Comment by tigrannajaryan** [comment]
> @tigrannajaryan @pjanotti @rghetia Nice, design is generic. I would like to implement this feature. If you have not started implementing it yet, let me know if you would want me to update 
this PR with the composite-policy Or Open a new PR ?

@vikrambe It would be great if you implemented this. I have a draft code implementing this  composite policy design that I wrote a while ago for OpenCensus (but never merged) here https://github.com/census-instrumentation/opencensus-service/compare/master...tigrannajaryan:feature/tigran/composite-policy?expand=1

Feel free to take it and use it. I would prefer the implementation to look similar to this. I would advise to create a new PR.


> w.r.t tail-sampling : Do we have a feature/design doc that solves tail-sampling across multiple otel-collector instances ?
> 
> **UseCase** : In order to handle high TPS, we would need multiple instances of otel collector behind a LoadBalancer. In this case, spans belonging to a trace are not guaranteed to land-up in the same collector instance. As the tail sampling works on the spans received by a single instance, sampling decision may potentially be flawed.
> 
> Solution : Enhance Opentelemetry collector to use distributed cache, so that OtelCollectors have a single view of the cache.

We don't have a design for this and we definitely would want to have a design doc and discussion for this before starting the implementation as this is a non-trivial feature.

**Comment by tigrannajaryan** [comment]
> @tigrannajaryan Thanks. I will start working on this and create a new PR.

Please create a new issue which we can use for discussion if needed and reference in your PR and commit message.

**Comment by pjanotti** [comment]
Sorry, missed this thread earlier.

> @pjanotti what are your thoughts on this? It seems composite policy has a real-world use case.

I think this path is very good. @vikrambe thanks for stepping up to do this, my understanding is that there will be a new PR coming, so I will close this one (let me know if that is not the case).

**Pull request closed without merging**